_posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide.md

```diff
@@ -0,0 +1,164 @@
+
+```
+desiredReplicas = ceil[currentReplicas * ( currentMetricValue / desiredMetricValue )]
+```
+
+Where:
+- `currentReplicas` is the number of replicas the deployment is currently running.
+- `currentMetricValue` is the current value of the metric we are monitoring (e.g. CPU usage).
+- `desiredMetricValue` is the target value for the metric we are monitoring (e.g. 90% requested CPU usage).
+
+So let's see what I observed in Grafana that day.
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide/memory_cpu.png --alt Grafana plot showing memory and CPU usages %}
+</div>
+
+{:refdef: class="image-caption"}
+*Figure 1. Grafana plot showing memory and CPU usages*
+{: refdef}
+
+What we observed was that at 13:10, we had three pods running with memory usage around `1100 MB` and CPU usage less than `100m`. Both metrics appeared to be below the target values of 90% of 1350 MB for memory and 90% of 500m for CPU. So, why were there three pods running?
+
+## Debugging HPA
+
+Let's start by inspecting what has happened to our application step by step using our visualization tool using previous metrics. First, let's check the memory usage:
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide/memory_init.png --alt HPA Memory expected replicas %}
+</div>
+
+{:refdef: class="image-caption"}
+*Figure 2. HPA memory expected replica*
+{: refdef}
+
+The memory usage seems to be below the 90% so the number of replicas would be set to 1. Let's check the CPU usage now:
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide/cpu_init.png --alt HPA CPU expected replicas %}
+</div>
+
+{:refdef: class="image-caption"}
+*Figure 3. HPA CPU expected replica*
+{: refdef}
+
+
+At startup, the CPU usage exceeded the target value of `450m`. This means that the 'currentMetricValue / desiredMetricValue' ratio was greater than one, indicating that the autoscaler needed to scale up the replicas. But by how much? Let’s adjust the x-axis of the plot to display the number of replicas:
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide/cpu_stairs.png --alt HPA CPU expected replicas showing stairs pattern %}
+</div>
+
+{:refdef: class="image-caption"}
+*Figure 4. HPA CPU current replicas vs expected replicas*
+{: refdef}
+
+There you go! We can clearly see a pattern resembling infinite stairs, similar to $f(x) = x + 1$. This occurs because the 'currentMetricValue / desiredMetricValue' ratio is greater than 1, prompting the autoscaler to continuously increase the number of replicas until it reaches the maximum allowed. In this instance, maxReplicas was set to 3. Thus, we have identified the root of the problem!
```

Comment:
I find this section a bit confusing. You seem to indicate that since the ratio between current and target is 600/450=1.33, thendesiredReplicas=currentReplicas + 1, so it enters an infinite loop that is only broken by the maxReplicas=3.

However, that is only true if the CPU utilisation (600) maintains as you add more replicas). What I would expect is that by adding more replicas, you spread the load, so each individual replica usage goes down. Like with 1 replica you have 600m usage. Then with 2 maybe you have 300m each (probably more since there's some fixed usage per replica).

But I'll go back to this later on because for me, this was the most important takeaway from this post.

Reply:
Also, there's a term for this: https://mathworld.wolfram.com/StaircaseFunction.html

```suggestion
There you go! We can clearly see the staircase pattern, similar to $f(x) = x + 1$. This occurs because the 'currentMetricValue / desiredMetricValue' ratio is greater than 1, prompting the autoscaler to continuously increase the number of replicas until it reaches the maximum allowed. In this instance, maxReplicas was set to 3. Thus, we have identified the root of the problem!
```

Reply:
Yeah, in this case the CPU load mantains even adding more replicas because the load does not seem to be related with traffic but with some high intensive task on app startup. It is worth it to add content to clarify this, thanks!